controller/konnect/reconciler_konnectnetworkref.go

```diff
@@ -0,0 +1,149 @@
+package konnect
+
+import (
+	"context"
+	"fmt"
+
+	sdkkonnectcomp "github.com/Kong/sdk-konnect-go/models/components"
+	"github.com/samber/lo"
+	"github.com/samber/mo"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/gateway-operator/controller/konnect/constraints"
+	sdkops "github.com/kong/gateway-operator/controller/konnect/ops/sdk"
+	"github.com/kong/gateway-operator/controller/pkg/patch"
+	k8sutils "github.com/kong/gateway-operator/pkg/utils/kubernetes"
+
+	commonv1alpha1 "github.com/kong/kubernetes-configuration/api/common/v1alpha1"
+	konnectv1alpha1 "github.com/kong/kubernetes-configuration/api/konnect/v1alpha1"
+)
+
+func handleKonnectNetworkRef[T constraints.SupportedKonnectEntityType, TEnt constraints.EntityType[T]](
+	ctx context.Context,
+	cl client.Client,
+	ent TEnt,
+	sdk sdkops.SDKWrapper,
+) (ctrl.Result, error) {
+	networkRefs, ok := getKonnectNetworkRefs(ent).Get()
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	setInvalidWithMsg := func(msg string) {
+		_ = patch.SetStatusWithConditionIfDifferent(
+			ent,
+			konnectv1alpha1.KonnectNetworkRefsValidConditionType,
+			metav1.ConditionFalse,
+			konnectv1alpha1.KonnectNetworkRefsReasonInvalid,
+			msg,
+		)
+	}
+
+	for _, ref := range networkRefs {
+		switch ref.Type {
+		case commonv1alpha1.ObjectRefTypeNamespacedRef:
+			network := konnectv1alpha1.KonnectCloudGatewayNetwork{}
+			nn := types.NamespacedName{
+				Name: ref.NamespacedRef.Name,
+				// TODO: handle cross namespace refs
+				Namespace: ent.GetNamespace(),
+			}
+
+			err := cl.Get(ctx, nn, &network)
+			if err != nil {
+				setInvalidWithMsg(err.Error())
+				return ctrl.Result{}, ReferencedObjectDoesNotExist{
+					Reference: nn,
+					Err:       err,
+				}
+			}
+
+			if delTimestamp := network.GetDeletionTimestamp(); !delTimestamp.IsZero() {
+				return ctrl.Result{}, ReferencedObjectIsBeingDeleted{
+					Reference:         nn,
+					DeletionTimestamp: delTimestamp.Time,
+				}
+			}
+
+			// requeue it if referenced network is not programmed yet so we cannot do the following work.
+			cond, ok := k8sutils.GetCondition(konnectv1alpha1.KonnectEntityProgrammedConditionType, &network)
+			if !ok || cond.Status != metav1.ConditionTrue {
+				setInvalidWithMsg(fmt.Sprintf("Referenced KonnectCloudGatewayNetwork %s is not programmed yet", nn))
+				return ctrl.Result{Requeue: true}, ReferencedObjectIsInvalid{
+					Reference: nn.String(),
+					Msg:       "Referenced KonnectCloudGatewayNetwork is not programmed yet",
+				}
```

Comment:
Let's either return an error or a `Result`. Returning an error should requeue anyway so I suggest omitting the `Result`.

Reply:
The result is useful e.g. when we get a conflict while patching https://github.com/Kong/gateway-operator/blob/d67c339d0fd7113f4ecf25d19ac9d6dd7aa20004/controller/konnect/reconciler_konnectnetworkref.go#L127-L132 to not explicitly handle an error at the call site.

Reply:
Yes, I mean here that in this line there's no point in returning a result with `Requeue` set to `true` as we're returning an error already.

Reply:
A yes, I removed setting the requeue as per the other comments 👍 PTAL.